lib/ui/painting/canvas.cc

```diff
@@ -114,63 +119,84 @@ void Canvas::saveLayer(double left,
                        double bottom,
                        const Paint& paint,
                        const PaintData& paint_data) {
-  if (!canvas_) {
-    return;
-  }
   SkRect bounds = SkRect::MakeLTRB(left, top, right, bottom);
-  canvas_->saveLayer(&bounds, paint.paint());
+  if (display_list_recorder_) {
+    bool restore_with_paint =
+        paint.sync_to(builder(), kSaveLayerWithPaintFlags);
+    builder()->saveLayer(&bounds, restore_with_paint);
+  } else if (canvas_) {
+    SkPaint sk_paint;
+    canvas_->saveLayer(&bounds, paint.paint(sk_paint));
+  }
 }
 
 void Canvas::restore() {
-  if (!canvas_) {
-    return;
+  if (display_list_recorder_) {
+    builder()->restore();
+  } else if (canvas_) {
+    canvas_->restore();
   }
-  canvas_->restore();
 }
 
 int Canvas::getSaveCount() {
-  if (!canvas_) {
+  if (display_list_recorder_) {
+    return builder()->getSaveCount();
+  } else if (canvas_) {
+    return canvas_->getSaveCount();
+  } else {
     return 0;
   }
-  return canvas_->getSaveCount();
 }
 
 void Canvas::translate(double dx, double dy) {
-  if (!canvas_) {
-    return;
+  if (display_list_recorder_) {
+    builder()->translate(dx, dy);
+  } else if (canvas_) {
+    canvas_->translate(dx, dy);
   }
-  canvas_->translate(dx, dy);
 }
 
 void Canvas::scale(double sx, double sy) {
-  if (!canvas_) {
-    return;
+  if (display_list_recorder_) {
+    builder()->scale(sx, sy);
+  } else if (canvas_) {
+    canvas_->scale(sx, sy);
   }
-  canvas_->scale(sx, sy);
 }
 
 void Canvas::rotate(double radians) {
-  if (!canvas_) {
-    return;
+  if (display_list_recorder_) {
+    builder()->rotate(radians * 180.0 / M_PI);
```

Comment:
Why is the Builder API using degrees, and the Canvas API using radians?

Reply:
Builder code is building from an SkPicture replacement so it matches SkCanvas rather than ui.Canvas. It is somewhat arbitrary, though.